docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,295 @@
+This interface can be exposed directly to consumers or we can implement a message queue or secondary streaming service on top.
+Using gRPC will provide us with all of the regular advantages of gRPC and protobuf: versioning guarantees, client side code generation, and interoperability with the many gRPC plugins and auxillary services.
+
+Proceeding through a gRPC intermediate will provide additional overhead, in most cases this is not expected to be rate limiting but in
+instances where it is the developer can implement a more performant streaming mechanism for state listening.
+
+### Configuration
+We will provide detailed documentation on how to configure the state listeners and their external streaming services from within an app's `AppCreator`,
+using the provided `AppOptions`. We will add two methods to the `BaseApp` to enable this configuration:
+
+```go
+// SetCommitMultiStoreListeners sets the KVStore listeners for the provided StoreKey
+func (app *BaseApp) SetCommitMultiStoreListeners(key sdk.StoreKey, listeners []storeTypes.WriteListener) {
+	app.cms.SetListeners(key, listeners)
+}
+
+// SetCacheListening turns on or off listening at the cache layer
+func (app *BaseApp) SetCacheListening(listening bool) {
+	app.cms.CacheListening(listening)
+}
+```
+
+As a demonstration, we will implement the state watching features as part of SimApp.
+For example, the below is a very rudimentary integration of the state listening features into the SimApp `AppCreator` function:
+
+
+```go
+func NewSimApp(
+	logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest bool, skipUpgradeHeights map[int64]bool,
+	homePath string, invCheckPeriod uint, encodingConfig simappparams.EncodingConfig,
+	appOpts servertypes.AppOptions, baseAppOptions ...func(*baseapp.BaseApp),
+) *SimApp {
+
+	...
+
+	keys := sdk.NewKVStoreKeys(
+		authtypes.StoreKey, banktypes.StoreKey, stakingtypes.StoreKey,
+		minttypes.StoreKey, distrtypes.StoreKey, slashingtypes.StoreKey,
+		govtypes.StoreKey, paramstypes.StoreKey, ibchost.StoreKey, upgradetypes.StoreKey,
+		evidencetypes.StoreKey, ibctransfertypes.StoreKey, capabilitytypes.StoreKey,
+	)
+	tkeys := sdk.NewTransientStoreKeys(paramstypes.TStoreKey)
+	memKeys := sdk.NewMemoryStoreKeys(capabilitytypes.MemStoreKey)
+
+	// configure state listening capabilities using AppOptions
+	if cast.ToBool(appOpts.Get("simApp.listening")) {
+		writeDir := filepath.Clean(cast.ToString(appOpts.Get("simApp.listening.writeDir")))
+		for _, key := range keys {
+			loadListener(bApp, writeDir, key)
+		}
+		for _, key := range tkeys {
+			loadListener(bApp, writeDir, key)
+		}
+		for _, key := range memKeys {
+			loadListener(bApp, writeDir, key)
+		}
+		bApp.SetCacheListening(cast.ToBool(appOpts.Get("simApp.cacheListening")))
+	}
+	
+	...
+
+	return app
+}
+
+// loadListener creates and adds to the BaseApp a listener that writes out to a file in the provided write directory
```

Comment:
I think we'll need to be able to map from `store.listeners` to helper functions. i.e. in app construction we look at all the listeners in the config and know which helper function to call`.

Reply:
Makes sense, I'll spec out a helper function lookup table. This would best fit as part of the `baseapp` package?